File: mobile/src/services/fcm/native-fcm.ts
// FCM(Phase D)ブリッジ。Android のネイティブプラグイン(FcmPlugin)から登録トークンを取得する。
// Android 以外(iOS/デスクトップ)では利用不可。
import { invoke } from '@tauri-apps/api/core'

export function isNativeFcmAvailable(): boolean {
  return typeof navigator !== 'undefined' && /Android/i.test(navigator.userAgent)
}

/** FCM 登録トークンを取得。取得できなければ null(権限未許可・Play開発者サービス無し等)。 */
export async function getFcmToken(): Promise<string | null> {
  try {
    const r = await invoke<{ token: string }>('get_fcm_token')
    return r.token || null
  } catch {
    return null
  }
}

File: mobile/src-tauri/src/native_fcm.rs
//! FCM(Phase D)。Kotlin の FcmPlugin(FirebaseMessaging.getToken)を登録し、アプリの
//! コマンド get_fcm_token から run_mobile_plugin で呼ぶ。Android 以外はエラー(JS側で分岐)。

use serde::{Deserialize, Serialize};
use tauri::plugin::{Builder, TauriPlugin};
use tauri::{AppHandle, Runtime};

#[cfg(target_os = "android")]
use tauri::{plugin::PluginHandle, Manager};

#[cfg(target_os = "android")]
struct NativeFcm<R: Runtime>(PluginHandle<R>);

#[derive(Debug, Serialize, Deserialize)]
pub struct FcmToken {
    pub token: String,
}

#[tauri::command]
pub fn get_fcm_token<R: Runtime>(app: AppHandle<R>) -> Result<FcmToken, String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<NativeFcm<R>>();
        state
            .0
            .run_mobile_plugin::<FcmToken>("getToken", ())
            .map_err(|e| e.to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Err("FCM は Android のみ対応です".into())
    }
}

pub fn init<R: Runtime>() -> TauriPlugin<R> {
    Builder::new("nativefcm")
        .setup(|_app, _api| {
            #[cfg(target_os = "android")]
            {
                let handle = _api.register_android_plugin("com.itsherpa.ffreceipt", "FcmPlugin")?;
                _app.manage(NativeFcm(handle));
            }
            Ok(())
        })
        .build()
}
